README.md
Using solitude for anomaly detection
====================================

-   Author: Srikanth Komala Sheshachala

The R package [solitude](https://cran.r-project.org/package=solitude)
implements **Isolation forest**, an anomaly detection method introduced
by the paper Isolation based Anomaly Detection ([Liu, Ting and
Zhou](https://dl.acm.org/citation.cfm?id=2133363)). Isolation forest is
grown using [ranger](https://cran.r-project.org/package=ranger)[1]
package.

‘Isolation forest’ is a multivariate outlier detection technique for IID
data containing a mix of numeric and categorical variables.

Usage
-----

    library("solitude")

    data("humus", package = "mvoutlier")
    columns_required = setdiff(colnames(humus)
                               , c("Cond", "ID", "XCOO", "YCOO", "LOI")
                               )
    humus2 = humus[ , columns_required]
    dplyr::glimpse(humus2)

    ## Rows: 617
    ## Columns: 39
    ## $ Ag <dbl> 0.124, 0.130, 0.124, 0.309, 0.040, 0.421, 0.235, 0.043, 0.070, 0.0…
    ## $ Al <dbl> 1420, 7190, 1970, 6360, 630, 2060, 1200, 1510, 3380, 1320, 2420, 1…
    ## $ As <dbl> 0.743, 0.973, 1.100, 1.240, 0.533, 1.540, 0.836, 0.657, 1.750, 1.0…
    ## $ B  <dbl> 2.51, 3.45, 1.62, 1.59, 4.33, 1.93, 2.51, 4.46, 5.08, 2.10, 2.18, …
    ## $ Ba <dbl> 74.0, 65.2, 121.0, 63.5, 36.0, 111.0, 109.0, 24.5, 42.2, 51.4, 103…
    ## $ Be <dbl> 0.02, 0.08, 0.04, 0.07, 0.01, 0.07, 0.01, 0.05, 0.06, 0.01, 0.06, …
    ## $ Bi <dbl> 0.108, 0.084, 0.137, 0.135, 0.076, 0.176, 0.145, 0.075, 0.259, 0.1…
    ## $ Ca <dbl> 3900, 1900, 3240, 2470, 3500, 3340, 3010, 3090, 1760, 2810, 3790, …
    ## $ Cd <dbl> 0.298, 0.390, 0.295, 0.222, 0.214, 0.327, 0.294, 0.180, 0.295, 0.2…
    ## $ Co <dbl> 1.28, 3.37, 0.98, 3.91, 0.43, 4.61, 0.75, 0.55, 1.68, 1.12, 1.36, …
    ## $ Cr <dbl> 3.04, 2.02, 1.33, 17.00, 0.79, 3.78, 2.01, 1.63, 10.90, 1.71, 2.20…
    ## $ Cu <dbl> 10.90, 12.30, 5.59, 29.80, 6.23, 47.30, 8.66, 5.55, 10.00, 9.25, 5…
    ## $ Fe <dbl> 2050, 3170, 1100, 6180, 790, 3370, 1150, 1800, 4820, 970, 1570, 15…
    ## $ Hg <dbl> 0.155, 0.263, 0.168, 0.175, 0.197, 0.239, 0.233, 0.217, 0.238, 0.2…
    ## $ K  <dbl> 1100, 900, 1000, 900, 700, 1200, 1000, 600, 800, 1000, 700, 1100, …
    ## $ La <dbl> 2.3, 6.4, 1.7, 5.3, 1.2, 7.0, 1.5, 2.9, 6.3, 1.2, 1.9, 1.8, 1.2, 1…
    ## $ Mg <dbl> 820, 670, 790, 1180, 1770, 800, 660, 1750, 820, 570, 690, 690, 590…
    ## $ Mn <dbl> 202.0, 22.8, 70.3, 118.0, 24.4, 86.1, 1210.0, 16.7, 39.2, 115.0, 1…
    ## $ Mo <dbl> 0.662, 0.286, 0.166, 0.280, 0.248, 0.473, 0.185, 0.372, 0.277, 0.1…
    ## $ Na <dbl> 40, 120, 30, 150, 410, 60, 10, 320, 110, 50, 20, 50, 20, 70, 50, 3…
    ## $ Ni <dbl> 9.41, 14.80, 5.05, 59.50, 1.98, 78.00, 5.70, 2.19, 11.70, 13.20, 5…
    ## $ P  <dbl> 743, 1030, 922, 717, 795, 942, 1120, 790, 779, 745, 869, 1160, 968…
    ## $ Pb <dbl> 15.9, 13.9, 19.0, 16.3, 14.3, 21.7, 15.2, 11.6, 284.0, 14.1, 18.1,…
    ## $ Rb <dbl> 8.14, 2.82, 4.45, 8.13, 2.67, 5.23, 4.94, 2.33, 4.07, 5.35, 3.69, …
    ## $ S  <dbl> 1300, 1950, 1750, 965, 1860, 1410, 1880, 2210, 1260, 1250, 1350, 1…
    ## $ Sb <dbl> 0.120, 0.161, 0.250, 0.017, 0.169, 0.217, 0.189, 0.126, 0.809, 0.2…
    ## $ Sc <dbl> 0.6, 1.2, 0.4, 1.4, 0.2, 0.7, 0.4, 0.6, 1.1, 0.3, 0.7, 0.4, 0.6, 0…
    ## $ Si <dbl> 630, 640, 580, 690, 440, 530, 650, 490, 630, 560, 660, 740, 700, 6…
    ## $ Sr <dbl> 22.2, 34.4, 43.4, 29.2, 46.4, 67.9, 20.7, 59.1, 28.8, 23.2, 27.7, …
    ## $ Th <dbl> 0.413, 0.281, 0.246, 0.816, 0.250, 0.432, 0.371, 0.535, 1.130, 0.1…
    ## $ Tl <dbl> 0.081, 0.068, 0.077, 0.099, 0.060, 0.084, 0.088, 0.042, 0.084, 0.0…
    ## $ U  <dbl> 0.240, 0.120, 0.042, 0.163, 0.127, 0.112, 0.051, 0.265, 0.221, 0.0…
    ## $ V  <dbl> 6.79, 3.89, 2.86, 16.10, 1.63, 7.40, 3.55, 2.57, 13.90, 2.76, 3.37…
    ## $ Y  <dbl> 0.8, 2.4, 0.5, 1.9, 0.8, 1.6, 0.5, 1.8, 1.7, 0.4, 0.8, 0.7, 0.7, 0…
    ## $ Zn <dbl> 59.1, 18.1, 67.6, 43.3, 41.4, 46.0, 103.0, 21.5, 23.2, 38.5, 53.6,…
    ## $ C  <dbl> 39.9, 47.5, 44.2, 19.4, 45.8, 47.8, 45.7, 47.3, 29.5, 46.8, 44.3, …
    ## $ H  <dbl> 5.5, 6.8, 6.3, 3.0, 6.1, 5.8, 6.5, 6.4, 4.4, 6.5, 6.1, 6.3, 5.2, 5…
    ## $ N  <dbl> 1.2, 1.5, 1.5, 0.7, 1.5, 1.3, 1.7, 1.5, 0.7, 1.0, 1.1, 1.6, 1.0, 1…
    ## $ pH <dbl> 3.9, 4.1, 3.8, 4.0, 4.1, 3.7, 4.0, 4.2, 4.0, 3.6, 3.7, 3.9, 3.8, 3…

    set.seed(1)
    index = sample(ceiling(nrow(humus2) * 0.5))

    # initiate an isolation forest
    iso = isolationForest$new(sample_size = length(index))

    # fit for attrition data
    iso$fit(dataset = humus2)

    ## Building Isolation Forest ... done
    ## Computing depth of terminal nodes ... done

    # Obtain anomaly scores on train data
    iso$scores

    ##       id average_depth anomaly_score
    ##   1:   1      21.04908     0.2531709
    ##   2:   2      17.37555     0.3217595
    ##   3:   3      22.20352     0.2347977
    ##   4:   4      14.43256     0.3898910
    ##   5:   5      17.98929     0.3091266
    ##  ---                                
    ## 613: 613      10.20077     0.5139055
    ## 614: 614      14.64205     0.3845970
    ## 615: 615      10.22036     0.5132491
    ## 616: 616      16.33663     0.3443316
    ## 617: 617      20.60641     0.2605914

    # predict outliers on test or unseen data
    iso$predict(humus2[-index, ]) # scores for new data (50% sample)

    ##       id average_depth anomaly_score
    ##   1:   1      19.08789     0.2877392
    ##   2:   2      11.84070     0.4617464
    ##   3:   3      16.77533     0.3346131
    ##   4:   4      20.32612     0.2654021
    ##   5:   5      20.86004     0.2563136
    ##  ---                                
    ## 304: 304      10.20077     0.5139055
    ## 305: 305      14.64205     0.3845970
    ## 306: 306      10.22036     0.5132491
    ## 307: 307      16.33663     0.3443316
    ## 308: 308      20.60641     0.2605914

Anomaly detection
-----------------

The paper suggests the following: If the score is closer to 1 for a some
observations, they are likely outliers. If the score for all
observations hover around 0.5, there might not be outliers at all.

By observing the quantiles, we might arrive at the a threshold on the
anomaly scores and investigate the outlier suspects.

    # quantiles of anomaly scores (of train data)
    quantile(iso$scores$anomaly_score
             , probs = seq(0.5, 1, length.out = 11)
             )

    ##       50%       55%       60%       65%       70%       75%       80%       85% 
    ## 0.2885520 0.2968820 0.3047881 0.3157465 0.3248082 0.3349006 0.3503038 0.3676153 
    ##       90%       95%      100% 
    ## 0.4148054 0.4684593 0.7180871

The understanding of *why is an observation an anomaly* might require a
combination of domain understanding and techniques like lime (Local
Interpretable Model-agnostic Explanations), Rule based systems etc

Installation
------------

    install.packages("solitude")                  # CRAN version
    devtools::install_github("talegari/solitude") # dev version

------------------------------------------------------------------------

[1] Wright MN, Ziegler A (2017). “ranger: A Fast Implementation of
Random Forests for High Dimensional Data in C++ and R.” Journal of
Statistical Software, 77(1), 1–17. doi: 10.18637/jss.v077.i01


NEWS.md
# solitude 1.0.0

- API is finalized.

# solitude 0.2.1

- added 'mtry' argument for isolationForest method
- minor optimizations in 'terminalNodesDepth' fuction

# solitude 0.2.0

- Changes in API: `isolationForest` is now a R6 class with new(), fit() and predict() methods.
- Exported `terminalNodesDepth` function to obtain the depth of terminal nodes of all trees in a ranger forest. 

# solitude 0.1.3

- Changed defaults to avoid bagging in extremly randomized trees (see https://stats.stackexchange.com/questions/406666/what-exactly-is-the-extratrees-option-in-ranger)

- Using `future_lapply` in favour of `furrr::furrr_map`


R/isolationForest_r6.R
#' @name isolationForest
#' @title Fit an Isolation Forest
#' @description 'solitude' class implements the isolation forest method
#'   introduced by paper Isolation based Anomaly Detection (Liu, Ting and Zhou
#'   <doi:10.1145/2133360.2133363>). The extremely randomized trees (extratrees)
#'   required to build the isolation forest is grown using
#'   \code{\link[ranger]{ranger}} function from \pkg{ranger} package.
#' @section Design: \code{$new()} initiates a new 'solitude' object. The
#'   possible arguments are:
#'
#'   \itemize{
#'
#'   \item \code{num_trees}: (positive integer, default = 100) Number of trees
#'   to be built in the forest
#'
#'   \item \code{sample_fraction}: ((0, 1], default = 1) Fraction of the dataset
#'   to be sampled or bootstrapped per tree. See 'sample.fraction' argument in
#'   \code{\link[ranger]{ranger}}
#'
#'   \item \code{replace}: (boolean, default = FALSE) Whether the sample of
#'   observations for each tree should be chosen with replacement. See 'replace'
#'   argument in \code{\link[ranger]{ranger}}
#'
#'   \item \code{seed}: (positive integer, default = 101) Random seed for the
#'   forest
#'
#'   \item \code{nproc}: (a positive integer, default: one less than maximum
#'   number of scores available) Number of parallel threads to be used by ranger
#'
#'   \item \code{respect_unordered_factors}: (string, default: "partition") See
#'   'respect.unordered.factors' argument in \code{\link[ranger]{ranger}}
#'
#'   }
#'
#'   \code{$fit()} fits a isolation forest for the given dataframe, computes
#'   depths of terminal nodes of each tree and stores the anomaly scores and
#'   average depth values in \code{$scores} object as a data.table
#'
#'   \code{$predict()} returns anomaly scores for a new data as a data.table
#'
#' @section Details:
#'
#'   \itemize{
#'
#'   \item Parallelization: \code{\link[ranger]{ranger}} is parallelized and by
#'   default uses all cores but one. The process of obtaining depths of terminal
#'   nodes (which is executed with \code{$fit()} is called) may be parallelized
#'   separately by setting up a \pkg{future} backend.
#'
#'   }
#'
#' @examples
#' data("humus", package = "mvoutlier")
#' columns_required = setdiff(colnames(humus)
#'                            , c("Cond", "ID", "XCOO", "YCOO", "LOI")
#'                            )
#' humus2 = humus[ , columns_required]
#' set.seed(1)
#' index = sample(ceiling(nrow(humus2) * 0.5))
#' isf = isolationForest$new()  # initiate
#' isf$fit(humus2[index, ])     # fit on 80% data
#' isf$scores                   # obtain anomaly scores
#'
#' # scores closer to 1 might indicate outliers
#' plot(density(isf$scores$anomaly_score))
#'
#' isf$predict(humus2[-index, ]) # scores for new data
#' @export

isolationForest = R6::R6Class(
  "solitude"
  ,
  # Implementation:
  # 'solitude' class implements the isolation forest method defined by
  # Liu, Fei Tony, Ting, Kai Ming and Zhou, Zhi-Hua.
  # "Isolation-based anomaly detection." ACM Transactions on Knowledge Discovery   # from Data (T KDD). <doi:10.1145/2133360.2133363>
  #
  # Plan:
  # 1. Build 'extratrees' forest using ranger.
  # 2. Obtain terminal node depths by parsing each tree.
  # 3. Compute anomaly scores.
  #
  # Design:
  # 1. 'initialize' method sets the params required for forest growth.
  # 2. Calling 'fit' on some dataset does these things:
  #    - Grow 'extratrees' forest.
  #    - Compute and store the depth of each terminal node in every tree.
  #    - Compute depth or pathlength for each observation.
  #    - Compute anomaly score for each observation and store it in 'scores'.
  # 3. 'predict' method computes anomaly scores on new data.
  #
  public = list(

    num_trees                   = NULL
    , replace                   = NULL
    , seed                      = NULL
    , nproc                     = NULL
    , respect_unordered_factors = NULL
    , sample_fraction           = NULL
    , scores                    = NULL
    , status                    = "not_initialized"
    ,
    # intialize arguments required for fitting extratrees via ranger
    initialize = function(
      num_trees                   = 100
      , replace                   = FALSE
      , sample_fraction           = 1
      , respect_unordered_factors = "partition"
      , seed                      = 101
      , nproc                     = parallel::detectCores() -1
      ){

      stopifnot(is_integerish(num_trees) && length(num_trees) == 1)
      stopifnot(0 < num_trees)
      stopifnot(is.logical(replace) && length(replace) == 1)
      stopifnot(is_integerish(seed) && length(seed) == 1)
      stopifnot(is_integerish(nproc) && length(nproc == 1) && nproc >= 1)
      stopifnot(is.character(respect_unordered_factors) &&
                  length(respect_unordered_factors) == 1)
      stopifnot(sample_fraction <= 1 && sample_fraction > 0)

      self$num_trees                 = num_trees
      self$replace                   = replace
      self$seed                      = seed
      self$nproc                     = nproc
      self$respect_unordered_factors = respect_unordered_factors
      self$sample_fraction           = sample_fraction

      self$status = "not_trained"
    }
    ,
    fit = function(dataset){

      # create new fit
      if(self$status == "trained"){
        self$status = "not_trained"
        warning("Retraining ... all previous training results will be lost")
      }

      # create a new 'y' column with jumbled 1:n
      columnNames  = colnames(dataset)
      nr           = nrow(dataset)
      nc           = ncol(dataset)
      responseName = columnNames[[1]]
      while (deparse(substitute(responseName)) %in% columnNames) {
        responseName = sample(c(letters, LETTERS), 20, replace = TRUE)
      }
      set.seed(self$seed + 1)
      dataset[[deparse(substitute(responseName))]] = sample(nrow(dataset))

      # build a extratrees forest
      message("Building Isolation Forest ... ", appendLF = FALSE)
      private$forest = ranger::ranger(
        dependent.variable.name     = deparse(substitute(responseName))
        , data                      = dataset
        , mtry                      = nc
        , min.node.size             = 1L
        , splitrule                 = "extratrees"
        , num.random.splits         = 1L
        , num.trees                 = self$num_trees
        , replace                   = self$replace
        , sample.fraction           = self$sample_fraction
        , respect.unordered.factors = self$respect_unordered_factors
        , num.threads               = self$nproc
        , oob.error                 = FALSE
        , seed                      = self$seed
        )
      message("done")

      # compute terminal nodes depth
      message("Computing depth of terminal nodes ... ", appendLF = FALSE)
      private$terminal_nodes_depth = terminalNodesDepth(private$forest)
      message("done")

      # set phi -- sample size used for tree building
      private$phi = floor(self$sample_fraction * nr)

      # predict anomaly scores for data used for fitting
      self$scores = self$predict(dataset)

      # update train status
      self$status = "trained"
    }
    ,
    predict = function(data){

      tnm = stats::predict(private$forest
                           , data
                           , type        = "terminalNodes"
                           , num.threads = self$nproc
                           )[["predictions"]]

      tnm = data.table::as.data.table(tnm)
      data.table::setnames(tnm, colnames(tnm), as.character(1:ncol(tnm)))
      tnm[, id := .I]
      tnm = data.table::melt(tnm
                             , id.vars         = "id"
                             , variable.name   = "id_tree"
                             , value           = "id_node"
                             , variable.factor = FALSE
                             )
      id_tree = NULL
      id_node = NULL

      tnm[, id_tree := as.integer(id_tree)]
      tnm[, id_node := as.integer(id_node)]

      obs_depth = merge(tnm
                        , private$terminal_nodes_depth
                        , by = c("id_tree", "id_node")
                        )

      # create extra path length when terminal nodes are not singletons
      path_length_extend = tnm[ , .N, c("id_tree", "id_node")]
      path_length_extend[
          , extend := vapply(N, private$pathLengthNormalizer, numeric(1))
          ]

      # extend length depending on terminal node
      obs_depth = merge(obs_depth
                        , path_length_extend
                        , by = c("id_tree", "id_node")
                        )
      obs_depth[ , depth := depth + extend]
      obs_depth[ , c("N", "extend") := NULL]

      average_depth = NULL
      depth         = NULL
      id            = NULL
      anomaly_score = NULL

      scores = obs_depth[ , .(average_depth = mean(depth)
                              , median_depth = median(depth)
                              )
                          , by = id][order(id)]
      scores[, anomaly_score :=
               private$computeAnomaly(average_depth, private$phi)]
      data.table::setorder(scores, -anomaly_score)

      return(scores[])
    }
  )
  ,
  private = list(

    forest                       = NULL
    , terminal_nodes_depth       = NULL
    , phi                        = NULL
    ,
    pathLengthNormalizer = function(phi){

      res = 0

      if(phi == 2){
        res = 1
      }
      if(phi > 2){
        res = (2 * private$harmonic(phi - 1)) - (2 * (phi - 1)/phi)
      }

      return(res)

    }
    ,
    computeAnomaly = function(depth, data_size){

      den = private$pathLengthNormalizer(data_size)
      return(2^( -( depth / den ) ))
    }
    ,
    harmonic = function(n){
      ifelse(n > 11, log(n) + 0.577216, sum(1/seq(1,n)))
    }
  )
)
